mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/task/SolveMailboxInconsistenciesServiceTest.java

```diff
@@ -0,0 +1,527 @@
+    @Test
+    void fixMailboxInconsistenciesShouldNotAlterStateWhenTwoEntriesWithSamePath() {
+        // Both mailbox merge is required
+        Mailbox mailbox2 = new Mailbox(MAILBOX_PATH, UID_VALIDITY_2, CASSANDRA_ID_2);
+
+        mailboxDAO.save(MAILBOX).block();
+        mailboxPathV2DAO.save(MAILBOX_PATH, CASSANDRA_ID_2).block();
+        mailboxDAO.save(mailbox2).block();
+
+        testee.fixMailboxInconsistencies(new Context()).block();
+
+        SoftAssertions.assertSoftly(softly -> {
+            softly.assertThat(mailboxDAO.retrieveAllMailboxes().collectList().block())
+                .containsExactlyInAnyOrder(MAILBOX, mailbox2);
+            softly.assertThat(mailboxPathV2DAO.listAll().collectList().block())
+                .containsExactlyInAnyOrder(
+                    new CassandraIdAndPath(CASSANDRA_ID_2, MAILBOX_PATH));
+        });
+    }
+
+    @Test
+    void concurrentCreateShouldNotBeConsideredAsAnInconsistency(CassandraCluster cassandra) throws Exception {
+        Barrier barrier = new Barrier();
+
+        cassandra.getConf()
+            .awaitOn(barrier)
+            .whenBoundStatementStartsWith("INSERT INTO mailbox ")
+            .times(TRY_COUNT_BEFORE_FAILURE)
+            .setExecutionHook();
+
+        // Start create a mailbox. Path registration entry will be created.
+        // However we instrument the driver to block mailbox entry until barrier is released
+        Mono<Mailbox> createMailbox = Mono.fromCallable(() -> mapper.create(MAILBOX_PATH, UID_VALIDITY_1))
+            .cache();
+        createMailbox.subscribeOn(Schedulers.elastic())
+            .subscribe();
+        barrier.awaitCaller();
+
+        // Start fixing inconsistencies
+        Context context = new Context();
+        ConcurrentTestRunner runner = ConcurrentTestRunner.builder()
+            .operation((a, b) -> testee.fixMailboxInconsistencies(context).block())
+            .threadCount(1)
+            .operationCount(1)
+            .run();
+
+        // Let the 'fix inconsistencies' make some progress...
+        Thread.sleep(GRACE_PERIOD_MILLIS / 2);
+        barrier.releaseCaller();
+
+        runner.awaitTermination(Duration.ofMinutes(1));
+        SoftAssertions.assertSoftly(softly -> {
+            // Fail on concurrent modification
+            softly.assertThat(context)
+                .isEqualTo(Context.builder()
+                    .processedMailboxPathEntries(1)
+                    .errors(1)
+                    .build());
+
+            // Don't alter DB state
+            Mailbox mailbox = createMailbox.block();
+            CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
+            MailboxAssertingTool.softly(softly)
+                .assertThat(mailboxDAO.retrieveMailbox(mailboxId).block())
+                .isEqualTo(mailbox);
+            assertThat(mailboxPathV2DAO.retrieveId(MAILBOX_PATH).block())
+                .isEqualTo(new CassandraIdAndPath(mailboxId, MAILBOX_PATH));
+        });
```

Comment:
How does it solve this scenario?
```
Server 1                                  Server 2

                                          create mailboxPathV2
solveInconsistencies()

is mailboxPathV2 alone?
                                          create mailbox        
delete mailboxPathV2
```
In this case (with just a slow Server 2) you end with:
- a bad response to the user stating that mailbox has been successfully created
- a mailbox created without a mailboxPath, breaking all your assumptions about consistency

Reply:
Well, the same way ;-)

Server 2 will not take more than 10 seconds to complete mailbox creation so we can infer server 1 will not wipe it out.

Reply:
I see several cases where this can easily happen:
- server 2 is really slow, not waiting for Cassandra. Then the 2xCassandra timeout rule has no real meaning
- don't you ever see p99>10s for a create mailbox?
- you just merge https://github.com/linagora/james-project/pull/3119 . How can you expect now that write to Cassandra are time bounded?

And I don't speak of harder scenarios where the users create / delete / recreate the mailbox. Nor about rename...

Anyway I do think this kind of problem can occur. How do you plan to solve them?

Reply:
>  Then the 2xCassandra timeout rule has no real meaning

> don't you ever see p99>10s for a create mailbox?

Increasing it a bit (eg: one minute) allows to be safer then.

p99 includes other latencies (http, thread allocation on james side before coming in mailbox/cassandra, mailbox provisionning checks, etc...)

I agree to choose a high value here (one minute?)

> you just merge #3119 . How can you expect now that write to Cassandra are time bounded?

As every write in there in time bounded, and the delay between them is as well, then it's just about choosing the long enough wait time. (IMO that's a detail)

> And I don't speak of harder scenarios where the users create / delete / recreate the mailbox. Nor about rename...

That look fun to test!

Hints: if the table status is not "exactly" the same we will not do any side effect and report a concurrent modification. The only issue here is *if we do a dirty read on the create during the diagnostic phase* and then *also do a dirty read upon delete (ie) on the validate phase*. That's very unlikely and IMO is not an issue here.

> How do you plan to solve them?

We cannot solve them.

We can reduce their occurrence probability.

Reply:
Can you define "exactly" please?

> We cannot solve them.

Before we had known inconsistencies. Now we have potential (rare) unknown inconsistencies, and lying to the client. I'm still not sure it's a good move.

Reply:
> Can you define "exactly" please?

Here it is used as semantic sugar.

Read: `if the table status is not [...] the same we will not do any side effect and report a concurrent modification.`

>> We cannot solve them.
> Before we had known inconsistencies. Now we have potential (rare) unknown inconsistencies, and lying to the client. I'm still not sure it's a good move.

After running such a task we've got:

 - Potentially fixed inconsistencies => that's good!
 - Detected conflicts requiring admin actions => at least they know ;-)
 - Detected concurrent actions leading to warnings => no harm
 - Undetected concurrent actions => they are caused by massive delays on server side. If we choose a long enough timeout, we can assume the user thinks things are failed. EG: 503 gateway. Or will do speculative retries.

Reply:
> Here it is used as semantic sugar.

> Read: if the table status is not [...] the same we will not do any side effect and report a concurrent modification.

I understood this. My question was: do you compare the full row, including the time the row was added (if available), or just the James data (mailboxPath / mailboxId)?

Reply:
So far James data

Are you looking for `ABA` abberacy?

Reply:
There's something I don't agree with:

> Undetected concurrent actions => they are caused by massive delays on server side. If we choose a long enough timeout, we can assume the user thinks things are failed. EG: 503 gateway. Or will do speculative retries.

It's very optimistic to come with a list of causes for such problems. I mean, my brain is obviously too small to cover all concurrency situations that can happen.

So let's consider you are not able to achieve that kind of analysis perfectly, what can we do to:

* be able to audit this task outcomes during first weeks to gain confidence about it
* be able to audit a concurrency issue after we have a consistency issue reported
* be able to fix a mistake this task did

Reply:
The output of the task details should allow monitoring and answering such questions.

Webadmin create mailbox endpoint should allow fixing issues.

Reply:
Re-speaking about this Hence Iwith @rouazana :

 - We need a **dryRun**. Running it on a production server, we can monitor the evolution of the reported inconsistency. If we notice that some inconsistencies 'disappear' this is a stong evidence of concurrency issues. Close monitoring of this will better document the level of harmfulness of this task.

Hence I propose:
 - 1. to merge this work
 - 2. to rebase & merge https://github.com/linagora/james-project/pull/3114 the webadmin endpoint - document it as experimental and explain the exposure to concurrency issues is currently undocumented.
 - 3. contribute a **dryRun** option.
 - 4. Ensure a nightly dryRun report from a platform with meaningful data and interaction end up on the private linagora james ML for review.

Reply:
I agree with you we really need this work to solves some inconsistencies harming some customers.

I would just like that we remove the delays for this PR:
- delays are solving a specific issue here, but there is no way to know they are not worsening other use cases
- with (ii.) we ask admins to avoid concurrency (for example by running by night or offline), but adding delays will complexify this task

In the future I would really like that we come with a better data model for mailboxes, allowing to have one source of truth and only projection to regenerate/fix when needed.